### ISSUE TITLE
Error calling gRPC Service Invocation with .NET

### ISSUE THREAD

**Issue by halspang** [dapr/dapr#7012]
## In what area(s)?
/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## What version of Dapr?

1.12.0-rc.x

## Expected Behavior

gRPC invocation should work as it did in 1.11.

## Actual Behavior

The .NET gRPC invocation example (non-proxy) fails to complete a request. The issue appears to be happening with the `Unpack` of an `Any` field in the proto's message. 

```
== APP == info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
== APP ==       Request starting HTTP/2 POST http://127.0.0.1:5050/dapr.proto.runtime.v1.AppCallback/OnInvoke application/grpc -
== APP == info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
== APP ==       Executing endpoint 'gRPC - /dapr.proto.runtime.v1.AppCallback/OnInvoke'
== APP == fail: Grpc.AspNetCore.Server.ServerCallHandler[6]
== APP ==       Error when executing service method 'OnInvoke'.
== APP ==       Google.Protobuf.InvalidProtocolBufferException: Full type name for Transaction is Transaction; Any message's type url is
== APP ==          at Google.Protobuf.WellKnownTypes.Any.Unpack[T]()
== APP ==          at GrpcServiceSample.BankingService.OnInvoke(InvokeRequest request, ServerCallContext context) in C:\Users\halspang\Workspace\dotnet-sdk\examples\AspNetCore\GrpcServiceSample\Services\BankingService.cs:line 72
== APP ==          at Grpc.Shared.Server.UnaryServerMethodInvoker`3.AwaitInvoker(Task`1 invokerTask, GrpcActivatorHandle`1 serviceHandle)
== APP ==          at Grpc.Shared.Server.UnaryServerMethodInvoker`3.AwaitInvoker(Task`1 invokerTask, GrpcActivatorHandle`1 serviceHandle)
== APP ==          at Grpc.AspNetCore.Server.Internal.CallHandlers.UnaryServerCallHandler`3.HandleCallAsyncCore(HttpContext httpContext, HttpContextServerCallContext serverCallContext)
== APP ==          at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerCallHandlerBase`3.<HandleCallAsync>g__AwaitHandleCall|8_0(HttpContextServerCallContext serverCallContext, Method`2 method, Task handleCall)
== APP == info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
== APP ==       Executed endpoint 'gRPC - /dapr.proto.runtime.v1.AppCallback/OnInvoke'
== APP == info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
== APP ==       Request finished HTTP/2 POST http://127.0.0.1:5050/dapr.proto.runtime.v1.AppCallback/OnInvoke application/grpc - - 200 0 application/grpc 119.7066ms
```

## Steps to Reproduce the Problem

Run the .NET [gRPC service invocation example](https://github.com/dapr/dotnet-sdk/blob/master/examples/Client/ServiceInvocation/README.md)

## Release Note

RELEASE NOTE: **FIX** Invocation error in gRPC service to service call

**Comment by halspang** [comment]
Note that this works with Dapr 1.11.4 with no changes in .NET.

**Comment by philliphoff** [comment]
Looking at this, too, just for the experience.  I believe I've verified:

 - The Any::Type property is set properly in the .NET client app and received by its Dapr sidecar
 - The .NET service app does not receive requests with the Any::Type property set and neither does its Dapr sidecar

This implies that the Any::Type property is getting lost in the Dapr sidecar to Dapr sidecar transition and not an issue with the .NET SDK itself

**Comment by artursouza** [comment]
I did a git bisect and it points this to be the offending commit. Working on a fix now.

```txt
$ git bisect bad
2814e6e9d95d861cac6ab4cafedecabcc5cbc98d is the first bad commit
commit 2814e6e9d95d861cac6ab4cafedecabcc5cbc98d
Author: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>
Date:   Wed Jul 12 15:08:50 2023 -0700

    Enable ServiceInvocationStreaming by default in builds (#6651)
    
    * Enable ServiceInvocationStreaming by default in builds
    
    Includes a fix for HTTP->gRPC service invocation
    
    Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>
    
    * 💄 the changes that matter
    
    Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>
    
    * Fixed issue with gRPC tracing
    
    Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>
    
    ---------
    
    Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

 Makefile                         |  2 +-
 pkg/channel/grpc/grpc_channel.go |  7 +++++
 pkg/diagnostics/grpc_tracing.go  | 60 ++++++++++++++++++++++++++++------------
 pkg/grpc/api_daprinternal.go     |  6 ++++
 4 files changed, 56 insertions(+), 19 deletions(-)
 ```

**Comment by artursouza** [comment]
Disabling `ServiceInvocationStreaming` from Makefile fixes the problem. So, it is something related to the streaming code.

**Comment by philliphoff** [comment]
Yes, 1.12 makes service invocation streaming the default whereas, in 1.11, it had to be enabled via feature flags. It looks like the issue predates 1.12, as you can reproduce the issue on 1.11.4 by explicitly enabling streaming.

The issue centers around the reconstruction of the request data (after being reset) but where `TypeUrl` is not restored:

https://github.com/dapr/dapr/blob/cfbc8182e4c629038b0b2a0cc43dd52bd1df2684/pkg/messaging/v1/invoke_method_request.go#L229

**Comment by mukundansundar** [comment]
Fix needed in 1.11 also ...

**Comment by ItalyPaleAle** [comment]
Closing as complete:

- Fixed in 1.12 branch.
- Fix for 1.11 branch is in #7018
